lib/delta_tables_generator/importer.rb

```diff
@@ -0,0 +1,191 @@
+        %i[goods_nomenclature_item_id
+           goods_nomenclature_sid
+           productline_suffix
+           end_line
+           delta_type
+           delta_date]
+      end
+
+      def where_condition(day: Date.current)
+        'true'
+      end
+
+      def integrate_element(row:, day: Date.current, is_end_line: nil, siblings: nil)
+        [
+          row[:goods_nomenclature_item_id],
+          row[:goods_nomenclature_sid],
+          row[:producline_suffix] || row[:productline_suffix] || '80',
+          is_end_line || end_line?(row: row, day: day, siblings: siblings),
+          delta_type,
+          day,
+        ]
+      end
+
+      def delta_type
+        'commodity'
+      end
+
+      def end_line?(row:, day: Date.current, siblings: [])
+        TimeMachine.at(day) do
+          item_id = row[:goods_nomenclature_item_id]
+          return false if chapter?(item_id) || heading?(item_id)
+
+          suffix = row[:producline_suffix] || row[:productline_suffix]
+          return false if suffix.present? && suffix != '80'
+
+          children = siblings&.any? ?
+                       find_children_from_siblings(row: row, siblings: siblings) :
+                       find_children(row: row, day: day)
+          children.none?
+        end
+      end
+
+      def integrate_and_find_children(row:, day: Date.current)
+        suffix = row[:producline_suffix] || row[:productline_suffix]
+        children = find_children(row: row, day: day)
+        is_end_line = suffix.nil? || suffix == '80' ? children.none? : false
+
+        [integrate_element(row: row, day: day, is_end_line: is_end_line)]
+          .concat(children.map { |child| integrate_element(row: child, day: day, siblings: children) })
+      end
+
+      def find_children_from_siblings(row:, siblings:)
+        item_id = row[:goods_nomenclature_item_id]
+        row_suffix = row[:producline_suffix] || row[:productline_suffix]
+
+        if chapter?(item_id)
+          children = siblings.select do |element|
+            /^#{item_id[0, 2]}/ =~ element[:goods_nomenclature_item_id]
+          end
+        elsif heading?(item_id)
+          children = siblings.select do |element|
+            /^#{item_id[0, 4]}/ =~ element[:goods_nomenclature_item_id]
+          end
+        else
+          goods_nomenclature = GoodsNomenclature.eager(:goods_nomenclature_indents)
+                                                .first(goods_nomenclature_sid: row[:goods_nomenclature_sid])
+          children = siblings.select do |element|
+            suffix = element[:producline_suffix] || element[:productline_suffix]
+            element[:number_indents] > goods_nomenclature.number_indents &&
+              (element[:goods_nomenclature_item_id] > item_id ||
+               (element[:goods_nomenclature_item_id] == item_id &&
+                suffix > row_suffix))
+          end
+        end
+
+        return children
+      end
+
+      def find_children(row:, day: Date.current)
```

Comment:
This method is quite heavy. Maybe move each sql statement in separate methods and break it in smaller chunks?

Reply:
True, that helps making it a bit more readable, thanks :)